In [1]:
#importing neccesaries libraries
import pandas as pd
import numpy as np
from datetime import datetime

# 1. Load the dataset


In [9]:
file_path = r"C:\Users\baliq\Downloads\MOCK_DATA (3).CSV"
#getting the number of rows
total_rows = sum(1 for _ in open(file_path)) - 1    # subtracting the header

#calculating 75% of the total rows\
rows_to_load = int(0.75 * total_rows)

#Reading only 75% of the data
df = pd.read_csv(file_path, nrows = rows_to_load)

df

Customer ID                Country  Total Purchases  \
0           1015  Palestinian Territory               12   
1           1029                  China               47   
2           1025                Nigeria               10   
3           1006         United Kingdom               48   
4           1024            Philippines               16   
..           ...                    ...              ...   
745         1024           South Africa               16   
746         1009                Nigeria               21   
747         1012                 Angola               29   
748         1003              Indonesia               30   
749         1014               Thailand                4   

                      Sign-up Date  
0    2022-04-03 21:42:29.000000000  
1    2016-01-08 03:13:12.000000000  
2    2023-02-20 19:06:14.000000000  
3    2022-05-26 12:46:35.000000000  
4    2016-12-13 03:44:04.000000000  
..                             ...  
745  2023-06-29 16:06:53.000000000  
746  2016-08-16 01:46:44.000000000  
747  2022-10-10 22:54:07.000000000  
748  2015-09-17 14:03:25.000000000  
749  2020-12-22 14:40:21.000000000  

[750 rows x 4 columns]

There are 750 rows and 4columns in this dataset

# 2. Data Cleaning


In [3]:
# Removing duplicates in the dataset
df_non_duplicate = df.drop_duplicates (keep = "first")
df_non_duplicate

Customer ID                Country  Total Purchases  \
0           1015  Palestinian Territory               12   
1           1029                  China               47   
2           1025                Nigeria               10   
3           1006         United Kingdom               48   
4           1024            Philippines               16   
..           ...                    ...              ...   
745         1024           South Africa               16   
746         1009                Nigeria               21   
747         1012                 Angola               29   
748         1003              Indonesia               30   
749         1014               Thailand                4   

                      Sign-up Date  
0    2022-04-03 21:42:29.000000000  
1    2016-01-08 03:13:12.000000000  
2    2023-02-20 19:06:14.000000000  
3    2022-05-26 12:46:35.000000000  
4    2016-12-13 03:44:04.000000000  
..                             ...  
745  2023-06-29 16:06:53.000000000  
746  2016-08-16 01:46:44.000000000  
747  2022-10-10 22:54:07.000000000  
748  2015-09-17 14:03:25.000000000  
749  2020-12-22 14:40:21.000000000  

[750 rows x 4 columns]

There are no duplicates in the dataset

In [30]:
# check for missing values
missing_values = df.isnull().sum()
print(missing_values)

Customer ID         0
Country             0
Total Purchases     0
Sign-up Date        0
Membership Years    0
Loyalty Score       0
dtype: int64


There are no missing values

In [4]:
df["Sign-up Date"] = pd.to_datetime (df["Sign-up Date"])   #Converting to proper date format

In [5]:
df

Customer ID                Country  Total Purchases        Sign-up Date
0           1015  Palestinian Territory               12 2022-04-03 21:42:29
1           1029                  China               47 2016-01-08 03:13:12
2           1025                Nigeria               10 2023-02-20 19:06:14
3           1006         United Kingdom               48 2022-05-26 12:46:35
4           1024            Philippines               16 2016-12-13 03:44:04
..           ...                    ...              ...                 ...
745         1024           South Africa               16 2023-06-29 16:06:53
746         1009                Nigeria               21 2016-08-16 01:46:44
747         1012                 Angola               29 2022-10-10 22:54:07
748         1003              Indonesia               30 2015-09-17 14:03:25
749         1014               Thailand                4 2020-12-22 14:40:21

[750 rows x 4 columns]

In [6]:
# replace all entries from the year 2020 with NaN.
df.loc[df["Sign-up Date"].dt.year == 2020, "Sign-up Date"] = np.NaN
df

Customer ID                Country  Total Purchases        Sign-up Date
0           1015  Palestinian Territory               12 2022-04-03 21:42:29
1           1029                  China               47 2016-01-08 03:13:12
2           1025                Nigeria               10 2023-02-20 19:06:14
3           1006         United Kingdom               48 2022-05-26 12:46:35
4           1024            Philippines               16 2016-12-13 03:44:04
..           ...                    ...              ...                 ...
745         1024           South Africa               16 2023-06-29 16:06:53
746         1009                Nigeria               21 2016-08-16 01:46:44
747         1012                 Angola               29 2022-10-10 22:54:07
748         1003              Indonesia               30 2015-09-17 14:03:25
749         1014               Thailand                4                 NaT

[750 rows x 4 columns]

In [7]:
#Drop all rows where "Total Purchases" is less than 5 and customer is from Canada.
df= df.drop(df [((df["Total Purchases"]< 5) & (df["Country"] == "Canada"))].index)
df

Customer ID                Country  Total Purchases        Sign-up Date
0           1015  Palestinian Territory               12 2022-04-03 21:42:29
1           1029                  China               47 2016-01-08 03:13:12
2           1025                Nigeria               10 2023-02-20 19:06:14
3           1006         United Kingdom               48 2022-05-26 12:46:35
4           1024            Philippines               16 2016-12-13 03:44:04
..           ...                    ...              ...                 ...
745         1024           South Africa               16 2023-06-29 16:06:53
746         1009                Nigeria               21 2016-08-16 01:46:44
747         1012                 Angola               29 2022-10-10 22:54:07
748         1003              Indonesia               30 2015-09-17 14:03:25
749         1014               Thailand                4                 NaT

[748 rows x 4 columns]

There are only 2 Total purchase from Canada that is less than 5

## 3. Feature Engineering

In [10]:
df["Sign-up Date"] = pd.to_datetime(df["Sign-up Date"])

# Calculate membership duration (in years)
current_year = pd.Timestamp.today().year
df["Membership Years"] = current_year - df["Sign-up Date"].dt.year

# Define function to calculate loyalty score
def assign_loyalty_score(row):
    if row["Total Purchases"] > 20:
        return 3
    elif 10 <= row["Total Purchases"] <= 20:
        return 2
    elif row["Total Purchases"] < 10:
        return 2 if row["Membership Years"] > 3 else 1

# Apply function to create new column
df["Loyalty Score"] = df.apply(assign_loyalty_score, axis=1)

print(df)

     Customer ID                Country  Total Purchases        Sign-up Date  \
0           1015  Palestinian Territory               12 2022-04-03 21:42:29   
1           1029                  China               47 2016-01-08 03:13:12   
2           1025                Nigeria               10 2023-02-20 19:06:14   
3           1006         United Kingdom               48 2022-05-26 12:46:35   
4           1024            Philippines               16 2016-12-13 03:44:04   
..           ...                    ...              ...                 ...   
745         1024           South Africa               16 2023-06-29 16:06:53   
746         1009                Nigeria               21 2016-08-16 01:46:44   
747         1012                 Angola               29 2022-10-10 22:54:07   
748         1003              Indonesia               30 2015-09-17 14:03:25   
749         1014               Thailand                4 2020-12-22 14:40:21   

     Membership Years  Loyalty Score  


## 4. Data Aggregation & Filtering

In [11]:
#Calculate avegrage and maximum purchase by country
Avg_purchases = df.groupby("Country")["Total Purchases"].mean().reset_index()
Max_purchases = df.groupby("Country")["Total Purchases"].max().reset_index()

print("Average Purchases per Country:")
print(Avg_purchases)

print("\nMax Purchases per Country:")
print(Max_purchases)

Average Purchases per Country:
         Country  Total Purchases
0    Afghanistan         8.000000
1        Albania        20.500000
2        Andorra        11.000000
3         Angola        15.000000
4      Argentina        24.692308
..           ...              ...
107      Vanuatu        22.000000
108    Venezuela        25.333333
109      Vietnam        24.333333
110        Yemen        10.000000
111     Zimbabwe         8.000000

[112 rows x 2 columns]

Max Purchases per Country:
         Country  Total Purchases
0    Afghanistan               20
1        Albania               38
2        Andorra               11
3         Angola               29
4      Argentina               45
..           ...              ...
107      Vanuatu               22
108    Venezuela               43
109      Vietnam               47
110        Yemen               15
111     Zimbabwe                8

[112 rows x 2 columns]


In [13]:
# Excluding country mexico
excluded_country = "Mexico"
filtered_df = df[df["Country"] != "Mexico"]

# Find the most common loyalty score per country
most_common_loyalty = filtered_df.groupby("Country")["Loyalty Score"].agg(lambda x: x.mode().iloc[0])

# Display result
print(most_common_loyalty)

Country
Afghanistan    2
Albania        2
Andorra        2
Angola         2
Argentina      3
              ..
Vanuatu        3
Venezuela      3
Vietnam        3
Yemen          2
Zimbabwe       2
Name: Loyalty Score, Length: 111, dtype: int64


In [29]:
df["Country"].value_counts()

Country
China          151
Indonesia       84
Russia          55
Philippines     38
Brazil          37
              ... 
Cameroon         1
Fiji             1
Cambodia         1
Uzbekistan       1
North Korea      1
Name: count, Length: 112, dtype: int64

## 5. Data Export & Report Generation

In [14]:
#Get today's date in the format DD-MM-YYYY
today_date = datetime.today().strftime("%d-%m-%Y")

# Generate filename in reverse order (YYYY_MM_DD)
filename = f"processed_data_{today_date.replace('-', '_')}.csv"

# Select the required columns
final_df = df[["Customer ID", "Loyalty Score", "Country", "Total Purchases"]]

# Shuffle the first 10 rows before displaying (without sorting)
shuffled_sample = final_df.sample(n=10, random_state=42)

# Save the DataFrame as a CSV
final_df.to_csv(filename, index=False)

# Print shuffled sample
print(shuffled_sample)

     Customer ID  Loyalty Score                Country  Total Purchases
506         1012              3  Palestinian Territory               46
357         1025              3               Ethiopia               37
133         1009              2                Nigeria               10
250         1007              3                Nigeria               43
299         1008              2                  Aruba               11
680         1027              3                 Gambia               39
336         1023              3                 France               38
155         1031              3                Finland               39
528         1011              3                 Brazil               34
736         1014              3            Philippines               22


## REPORT

The entire dataset contains 1000 rows and 4 columns, since loading 75% of the data gave 750 rows

There are no duplicate in the dataset

There are only 2 Total purchase from Canada that is less than 5

The most common Loyalty Scores are 3 and 2

When we excluded Mexico from the most common loyalty score per country, we had 111 unique countries which signifies that there 
are 112 unique countries in the dataset


